src/main/sensors/esc_sensor.c

```diff
@@ -1518,6 +1440,658 @@ static void apdSensorProcess(timeUs_t currentTimeUs)
+            const uint16_t *pal = tribParamAddrLen + i;
+            const uint8_t len = *pal & TRIB_PAL_LEN_MASK;
+            if ((*pal & TRIB_PAL_SYS_MASK) == 0) {
+                // schedule writes for dirty address ranges
+                if (memcmp(paramPayload + offset, paramUpdPayload + offset, len) != 0) {
+                    // set dirty bit
+                    tribDirtyParams |= (1U << i);
+                    // invalidate param
+                    tribInvalidParams |= (1 << i);
+                    // invalidate param payload - will be available again when all params again cached
+                    paramPayloadLength = 0;
+                }
+            }
+            offset += len;
+        }
+        return true;
+    }
+    else if (cmd == TRIB_PARAM_CMD_RESET) {
+        // reset ESC
+        tribResetEsc = true;
+        return true;
+    }
+    else {
+        // invalid command
+        return false;
+    }
+}
+
+static void tribBuildReq(uint8_t req, uint8_t addr, void *src, uint8_t len, uint16_t framePeriod, uint16_t frameTimeout)
+{
+    reqbuffer[0] = req;     // req
+    reqbuffer[1] = 0;       // addr
+    reqbuffer[2] = 0;
+    reqbuffer[3] = addr;
+    reqbuffer[4] = len;     // length
+    reqbuffer[5] = 0;
+    reqLength = TRIB_HEADER_LENGTH;
+    if (src != NULL) {
+        memcpy(reqbuffer + TRIB_HEADER_LENGTH, src, len);
+        reqLength += len;
+    }
+    rrfsmFramePeriod = framePeriod;
+    rrfsmFrameTimeout = frameTimeout;
+}
+
+static void tribInvalidateParams(void)
+{
+    tribInvalidParams = ~(~1U << (ARRAYLEN(tribParamAddrLen) - 1));
+}
+
+static uint8_t tribCalcParamBufferLength()
+{
+    uint8_t len = 0;
+    for (uint8_t j = 0; j < ARRAYLEN(tribParamAddrLen); j++)
+        len += (tribParamAddrLen[j] & TRIB_PAL_LEN_MASK);
+    return len;
+}
+
+static bool tribBuildNextParamReq(void)
+{
+    // pending reset request...
+    if (tribResetEsc) {
+        tribResetEsc = false;
+
+        tribUncSetup = TRIB_UNCSETUP_INACTIVE;
+        tribInvalidParams = 0;
+        tribDirtyParams = 0;
+        paramMspActive = false;
+
+        uint16_t reset = 0;
+        tribBuildReq(0xD0, 0x00, &reset, 2, TRIB_PARAM_FRAME_PERIOD, TRIB_PARAM_WRITE_TIMEOUT);
+        return true;
+    }
+
+    // ...or pending write request...
+    uint8_t offset = 0;
+    for (uint16_t *pal = tribParamAddrLen, dirtybits = tribDirtyParams; dirtybits != 0; pal++, dirtybits >>= 1) {
+        const uint8_t addr = (*pal & TRIB_PAL_ADDR_MASK) >> 8;
+        const uint8_t len = *pal & TRIB_PAL_LEN_MASK;
+        if ((dirtybits & 0x01) != 0) {
+            void *payload;
+            uint32_t iq22Payload[2];
+            if (addr == TRIB_PARAM_IQ22_ADDR) {
+                // convert scaled uint32 -> IQ22
+                const uint32_t q22 = 1 << 22;
+                uint32_t *pw = iq22Payload;
+                uint32_t *pp = (uint32_t*)(paramUpdPayload + offset);
+                *pw++ = (((double)*pp++) / 100 * q22);
```

Comment:
Why is double used here? First, STM32F7/F4 does not support double, so this would be enumlated.
Second. I am not sure why use double/float at all, as this is all integers?

Reply:
IQ22 float conversion required better precision than w/ integer only math.
test - expected is 150
    const uint32_t q22 = 1 << 22;
    uint32_t iqv = 0x00001893; // <= from ESC
    uint32_t d1= round(((double)iqv) * 100000 / q22);
    uint32_t d2 = ((uint32_t)iqv) * 100000 / q22;
    printf("\niq22> %d, %d", d1, d2);
output
    iq22> 150, 149

Reply:
You could do it correctly with integers, but it would look ugly.
Please use float instead. Double is trouble.

Reply:
Verifying use of float w/ unit tests prior to change...

Reply:
float passes tests - float replaces double